cmd/pinniped-proxy/src/main.rs

```diff
@@ -1,9 +1,40 @@
+use std::convert::Infallible;
+
+use anyhow::Result;
+use hyper::{Server, service::{make_service_fn, service_fn}};
 use structopt::StructOpt;
 
 // Ensure the root crate is aware of the child modules.
 mod cli;
+mod service;
 
-fn main() {
+#[tokio::main]
+async fn main() -> Result<()> {
     let opt = cli::Options::from_args();
-    println!("{:#?}", opt);
+
+    // For every incoming connection, we make a new hyper `Service` to handle
+    // all incoming HTTP requests on that connection. This is done by passing a
+    // closure to the hyper `make_service_fn` which returns our custom `make_svc`
+    // function that can be used for each connection. 
+    let make_svc = make_service_fn(|_conn| {
+        // The closure just returns an async block that runs a service to handle
+        // all requests for the connection.
+        async {
+            // `service_fn` is a helper from the hyper crate which converts a
+            // function that returns a Response into a `Service`.
+            Ok::<_, Infallible>(service_fn(service::proxy))
+        }
+    });
+
+    let addr = ([127, 0, 0, 1], opt.port).into();
+
+    let server = Server::bind(&addr).serve(make_svc);
+
+    println!("Listening on http://{}", addr);
+
+    // Run the server for ever. If it returns with an error, return the
+    // result, otherwise, if it completes, we return Ok.
+    server.await?;
+
+    Ok(())
```

Comment:
what is this `Ok(())` doing? I guess that if everything goes fine, it never executes?

Reply:
It's just returning an empty Ok result. If you look back on line 12 at the signature of the `main` function, you'll see it returns a `Result<()>` which means that the function will return one of the two Result enums - either Ok (with a matching empty type of `()`) or an error. It is much clearer when the result has a value, such as `Ok(some_number)`, but a main function either succeeds or it errors. Different types of errors may be returned, but if the main function succeeds, there's no value associated with it.